docs/tutorials/dataset_creation/custom_space_serialization.py

```diff
@@ -0,0 +1,203 @@
+# fmt: off
+"""
+Serializing a custom space
+=========================================
+"""
+# %%%
+# In this tutorial you'll learn how to serialize a custom Gym observation space and use that
+# to create a Minari dataset with :class:`minari.DataCollectorV0`. We'll use the
+# simple `CartPole-v1 <https://gymnasium.farama.org/environments/classic_control/cart_pole/>`_ environment,
+# modify the observation space, show how to serialize it, and create a Minari dataset.
+#
+# Serializing a custom space can be applied to both observation and action spaces.
+#
+# Let's get started by importing the required modules:
+
+import json
+from typing import Any, Dict, Sequence, Union
+
+# %%
+import gymnasium as gym
+import numpy as np
+from gymnasium.spaces import Space
+from numpy.typing import NDArray
+
+import minari
+from minari import DataCollectorV0
+from minari.data_collector.callbacks import StepDataCallback
+from minari.serialization import deserialize_space, serialize_space
+
+
+# %% [markdown]
+# First we'll initialize the CartPole environment and take a look at its observation and action space.
+
+# %%
+env = gym.make("CartPole-v1")
+
+print(f"Observation space: {env.observation_space}")
+print(f"Observation space: {env.action_space}")
+
+# %% [markdown]
+# Now we can create a new observation space that inherits from the `gym.Space <https://gymnasium.farama.org/api/spaces/#the-base-class>`_ class.
+
+# %%
+
+
+class CartPoleObservationSpace(Space):
+    def __init__(
+        self,
+        low: NDArray[Any],
+        high: NDArray[Any],
+        shape: Sequence[int],
+        dtype: type[np.floating[Any]] = np.float32,
+    ):
+        self.low = np.full(shape, low, dtype=dtype)
+        self.high = np.full(shape, high, dtype=dtype)
+        super().__init__(shape, dtype)
+
+    def sample(self) -> NDArray[Any]:
+        """Sample a random observation according to low/high boundaries"""
+        sample = np.empty(self.shape)
+        sample = self.np_random.uniform(low=self.low, high=self.high, size=self.shape)
+        return sample.astype(self.dtype)
+
+    def contains(self, x: Any) -> bool:
+        """Return boolean specifying if x is a valid member of this space"""
+        if not isinstance(x, np.ndarray):
+            try:
+                x = np.asarray(x, dtype=self.dtype)
+            except (ValueError, TypeError):
+                return False
+
+        return bool(
+            np.can_cast(x.dtype, self.dtype)
+            and x.shape == self.shape
+            and np.all(x >= self.low)
+            and np.all(x <= self.high)
+        )
```

Comment:
Yeah I thought about making a custom action space instead since that's a little simpler to implement (ex: maybe an action space that only allows moving left) but thought that might not get the idea across of a good example for when someone should create a custom space. I'll take a look at the `minigrid` environment and see if there's a modification that makes sense and is simple to implement.

Reply:
To clarify: what I meant is that the MissionSpace of minigrid is not serializable, so this can be a use case for the tutorial